New resource: azurerm_eventgrid_partner_configuration

gerrytan opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

New resource for Event Grid Partner Configuration

- [REST API: partner configuration](https://learn.microsoft.com/en-us/rest/api/eventgrid/controlplane/partner-configurations?view=rest-eventgrid-controlplane-2022-06-15)
- [MS doc: Event Grid](https://learn.microsoft.com/en-us/azure/event-grid/)

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->



## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

Acceptance test ran with command: `make acctests SERVICE='eventgrid' TESTARGS=' run="TestAccEventGridPartnerConfiguration_"' TESTTIMEOUT='120m'`

Test output: [link](https://microsoftapc-my.sharepoint.com/:u:/g/personal/tangerry_microsoft_com/Ed8BXT1NG1NMoa6Rp_i7bzYBI2pUhj09gMZr1oc_omd0mQ?e=kQwesw) (access restricted, please DM me on Slack)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_eventgrid_partner_configuration` - new resource [GH-13611]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
#13611 Note: please do not close the issue as there are other resources still yet to be implemented.


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

wyattfry commented on internal/services/eventgrid/registration.go:
```diff
@@ -29,12 +29,15 @@ func (r Registration) WebsiteCategories() []string {
 func (r Registration) DataSources() []sdk.DataSource {
-	return []sdk.DataSource{}
+	return []sdk.DataSource{
+		// todo: decide if this needs a data source
```
Is this still needed? If not, please remove the comment.
```suggestion
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+			if existing.Model == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", *id)
+			}
+
```
```suggestion
			if existing.Model.Properties == nil {
				return fmt.Errorf("retrieving %s: `model.Properties` was nil", *id)
			}
			payload := existing.Model
			if payload.Properties.PartnerAuthorization == nil {
				payload.Properties.PartnerAuthorization = &partnerconfigurations.PartnerAuthorization{}
			}
```

wyattfry replied:
Then replace instances of `existing.Model` with `payload` to conform more closely to the contribution guide.

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+				existing.Model.Tags = pointer.To(config.Tags)
+			}
+
+			if err := client.CreateOrUpdateThenPoll(ctx, *id, *existing.Model); err != nil {
```
```suggestion
			if err := client.CreateOrUpdateThenPoll(ctx, *id, *payload); err != nil {
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+			}
+
+			if metadata.ResourceData.HasChange("partner_authorization") {
+				existing.Model.Properties.PartnerAuthorization.AuthorizedPartnersList = expandAuthorizedPartnersList(&config.PartnerAuthorizations)
```
```suggestion
				existing.Model.Properties.PartnerAuthorization.AuthorizedPartnersList = expandAuthorizedPartnersList(config.PartnerAuthorizations)
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+				ResourceGroup: id.ResourceGroupName,
+			}
+
+			if model := resp.Model; model != nil {
+				state.DefaultMaximumExpirationTimeInDays = *model.Properties.PartnerAuthorization.DefaultMaximumExpirationTimeInDays
+				state.PartnerAuthorizations = flattenAuthorizedPartnersList(model.Properties.PartnerAuthorization.AuthorizedPartnersList)
+
+				if model.Tags != nil {
+					state.Tags = *model.Tags
+				}
+			}
```
Many of these values can possible be `nil`, so we can check them first to avoid a panic.
```suggestion
			if model := resp.Model; model != nil {
				if props := model.Properties; props != nil {
					if auth := props.PartnerAuthorization; auth != nil {
						state.DefaultMaximumExpirationTimeInDays = pointer.From(auth.DefaultMaximumExpirationTimeInDays)
						state.PartnerAuthorizations = flattenAuthorizedPartnersList(auth.AuthorizedPartnersList)
					}
				}
				
				state.Tags = pointer.From(model.Tags)
			}
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+
+	for _, partnerAuth := range *partner {
+		partnerAuthorizations = append(partnerAuthorizations, PartnerAuthorization{
+			PartnerName:                      *partnerAuth.PartnerName,
+			PartnerRegistrationImmutableId:   *partnerAuth.PartnerRegistrationImmutableId,
+			AuthorizationExpirationTimeInUtc: *partnerAuth.AuthorizationExpirationTimeInUtc,
```
```suggestion
			PartnerName:                      pointer.From(partnerAuth.PartnerName),
			PartnerRegistrationImmutableId:   pointer.From(partnerAuth.PartnerRegistrationImmutableId),
			AuthorizationExpirationTimeInUtc: pointer.From(partnerAuth.AuthorizationExpirationTimeInUtc),
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+func flattenAuthorizedPartnersList(partner *[]partnerconfigurations.Partner) []PartnerAuthorization {
+	partnerAuthorizations := []PartnerAuthorization{}
```
```suggestion
	partnerAuthorizations := make([]PartnerAuthorization, 0)
```
Literals result in `nil` if empty, and thus not allowing `append()`ing. Using `make` ensures an empty slice is returned that can accept appending.

gerrytan replied:
`append` is immune to nil, check: https://go.dev/play/p/DQHTjN7TVfG. But I have refactored this to use make just to be consistent with the other codes.

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource_test.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+func (r EventGridPartnerConfigurationsTestResource) requiresImport(data acceptance.TestData) string {
+	template := r.basic(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_eventgrid_partner_configuration" "import" {
+  resource_group_name = azurerm_resource_group.test.name
+}
+`, template)
```
```suggestion
	return fmt.Sprintf(`
%s

resource "azurerm_eventgrid_partner_configuration" "import" {
  resource_group_name = azurerm_resource_group.test.name
}
`, r.basic(data))
```

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource_test.go:
```diff
@@ -0,0 +1,132 @@
+	})
+}
+
+func TestAccEventGridPartnerConfiguration_complete(t *testing.T) {
```
Could you please add more steps so the update path is also tested? e.g. complete and then basic. This will ensure that the other properties in complete also work in create.

wyattfry commented on internal/services/eventgrid/eventgrid_partner_configuration_resource_test.go:
```diff
@@ -0,0 +1,132 @@
+    authorization_expiration_time_in_utc = "%s"
+  }
+
+  tags = { "foo" = "bar" }
```
```suggestion
  tags = {
    foo = "bar"
  }
```
Just to make it more consistent with other test configurations.

wyattfry commented on website/docs/r/eventgrid_partner_configuration.html.markdown:
```diff
@@ -0,0 +1,82 @@
+
+* `partner_registration_immutable_id` - (Required) The immutable id of the corresponding partner registration.
+
+* `authorization_expiration_time_in_utc` - (Optional) Expiration time of the partner authorization. If this timer expires, any request from this partner to create, update or delete resources in subscriber's context will fail. Value should be in RFC 3339 format in UTC time zone, for example: "2025-02-04T00:00:00Z". If not specified, the authorization will expire after `default_maximum_expiration_t ...
```
```suggestion
* `authorization_expiration_time_in_utc` - (Optional) Expiration time of the partner authorization. Value should be in RFC 3339 format in UTC time zone, for example: "2025-02-04T00:00:00Z".

-> **Note:** If the time from `authorization_expiration_time_in_utc` expires, any request from this partner to create, update or delete resources in the subscriber's context will fail. If not specified, the authorization will expire after `default_maximum_expiration_time_in_days`.
```

wyattfry commented on website/docs/r/eventgrid_partner_configuration.html.markdown:
```diff
@@ -0,0 +1,82 @@
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
+
+* `create` - (Defaults to 30 minutes) Used when creating the Event Grid Partner Configuration.
+* `read` - (Defaults to 30 minutes) Used when retrieving the Event Grid Partner Configuration.
```
```suggestion
* `read` - (Defaults to 5 minutes) Used when retrieving the Event Grid Partner Configuration.
```

gerrytan commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,270 @@
+			}
+
+			param := partnerconfigurations.PartnerConfiguration{
+				Location: pointer.To("global"),
```
Location is not configurable, only `"global"` is a valid value. Please refer to the create screen in portal below. Note that there no location input.

<img src="https://github.com/user-attachments/assets/220126f8-0b8a-4aba-ae93-ed9fa6b1d9d3" width=200 />

gerrytan commented:
Thx @wyattfry , I have addressed all the feedbacks (marked with 👍emoji) and all acctest passes (log attached below). Please give this another review.

- acctest cmd: `make acctests SERVICE='eventgrid' TESTARGS='-run="TestAccEventGridPartnerConfiguration_"' TESTTIMEOUT='120m'`
- output: [acctest log at commit 6d3a022ed5](https://microsoftapc-my.sharepoint.com/:u:/g/personal/tangerry_microsoft_com/EccVA2XjboNGgff0sZiL2jIBcmdvLETSWMQvmibkT4Cc4w?email=wyatt.fry%40hashicorp.com&e=xe8fiH)

stephybun commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+				}
+
+				if model.Tags != nil {
+					state.Tags = *model.Tags
```
```suggestion
					state.Tags = pointer.From(model.Tags)
```

gerrytan replied:
Fixed in cdabb53aed

stephybun commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+			}
+
+			if model := resp.Model; model != nil {
+				if model.Properties != nil && model.Properties.PartnerAuthorization != nil {
```
For consistency
```suggestion
				if props := model.Properties; props != nil && props.PartnerAuthorization != nil {
```

gerrytan replied:
Fixed in cdabb53aed

stephybun commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+
+			if model := resp.Model; model != nil {
+				if model.Properties != nil && model.Properties.PartnerAuthorization != nil {
+					state.DefaultMaximumExpirationTimeInDays = *model.Properties.PartnerAuthorization.DefaultMaximumExpirationTimeInDays
```
```suggestion
					state.DefaultMaximumExpirationTimeInDays = pointer.From(props.PartnerAuthorization.DefaultMaximumExpirationTimeInDays)
```

gerrytan replied:
Fixed in cdabb53aed

stephybun commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+			if model := resp.Model; model != nil {
+				if model.Properties != nil && model.Properties.PartnerAuthorization != nil {
+					state.DefaultMaximumExpirationTimeInDays = *model.Properties.PartnerAuthorization.DefaultMaximumExpirationTimeInDays
+					state.PartnerAuthorizations = flattenAuthorizedPartnersList(model.Properties.PartnerAuthorization.AuthorizedPartnersList)
```
```suggestion
					state.PartnerAuthorizations = flattenAuthorizedPartnersList(props.PartnerAuthorization.AuthorizedPartnersList)
```

gerrytan replied:
Fixed in 3d2c2ab512

gerrytan commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+		},
+		"partner_authorization": &schema.Schema{
+			Type:     pluginsdk.TypeList,
+			Optional: true,
```
@stephybun 'Partner Authorization' is optional. I've tested creating a partner config without partner auth fine in portal as per below screenshot:

![image](https://github.com/user-attachments/assets/26f6efdb-b96d-4ad6-824f-7e19cec80e28)

gerrytan commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func expandAuthorizedPartnersList(partnerAuthorization []PartnerAuthorization) *[]partnerconfigurations.Partner {
+	partners := []partnerconfigurations.Partner{}
```
Fixed in 57d73d4723

gerrytan commented on internal/services/eventgrid/eventgrid_partner_configuration_resource.go:
```diff
@@ -0,0 +1,278 @@
+
+			if payload.Properties.PartnerAuthorization == nil {
+				payload.Properties.PartnerAuthorization = &partnerconfigurations.PartnerAuthorization{}
+			}
```
Fixed in b0ee09a05c

gerrytan commented on website/docs/r/eventgrid_partner_configuration.html.markdown:
```diff
@@ -0,0 +1,84 @@
+
+# azurerm_eventgrid_partner_configuration
+
+Manages an Event Grid Partner Configuration
```
Fixed in faaa354470

gerrytan commented:
Thanks @stephybun, I have addressed all feedbacks, can you please have another look at this PR 🙏.

gerrytan commented:
acctest passes at latest commit: [acctests_TestAccEventGridPartnerConfiguration_2025-03-31T1523+1100_faaa354470.log](https://microsoftapc-my.sharepoint.com/:u:/g/personal/tangerry_microsoft_com/EYX7UiswhspNuE4AVmZpIg8BQka_ob-Fdu-GcQe4V9nKBQ?email=steph%40hashicorp.com&e=sP5ps3)
